app/models/concerns/generate_token.rb

```diff
@@ -0,0 +1,16 @@
+module GenerateToken
+  extend ActiveSupport::Concern
+
+  included do
+    def ensure_token
+      self.token = generate_hex(:token) unless token.present?
+    end
+
+    def generate_hex(column)
+      loop do
+        hex = SecureRandom.hex(32)
+        break hex unless self.class.where(column => hex).any?
```

Comment:
Rails has some built-in token generating methods. Could you look into using those? https://api.rubyonrails.org/classes/ActiveRecord/SecureToken/ClassMethods.html